### PR TITLE
Codechange: Use max_element to find best cargo (sub-)type in NewGRF var 42

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

Use of C-style arrays with memset, along with index array access and loops to find best element.

## Description

Replace C-style arrays / memset, with initialized std::array.
Replace the two loops to find best element with std::max_element.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/newgrf_engine.cpp [hunk 1]
```diff
@@ -435,19 +435,12 @@ static uint32_t VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *objec
 
 		case 0x42: { // Consist cargo information
 			if (!HasBit(v->grf_cache.cache_valid, NCVV_CONSIST_CARGO_INFORMATION)) {
-				const Vehicle *u;
+				std::array<uint8_t, NUM_CARGO> common_cargoes{};
+				std::array<uint8_t, UINT8_MAX + 1> common_subtypes{};
```
Comment (rubidium42) [inline src/newgrf_engine.cpp]: Maybe allocate when it's needed (~line 453)?

File: src/newgrf_engine.cpp [hunk 2]
```diff
@@ -458,30 +451,22 @@ static uint32_t VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *objec
...
-				uint common_cargo_best_amount = 0;
-				for (CargoID cargo = 0; cargo < NUM_CARGO; cargo++) {
-					if (common_cargoes[cargo] > common_cargo_best_amount) {
-						common_cargo_best_amount = common_cargoes[cargo];
-						common_cargo_type = cargo;
-					}
-				}
+				auto cargo_it = std::max_element(std::begin(common_cargoes), std::end(common_cargoes));
+				/* Return CT_INVALID if nothing is carried */
+				CargoID common_cargo_type = (*cargo_it == 0) ? (CargoID)CT_INVALID : static_cast<CargoID>(cargo_it - std::begin(common_cargoes));
```
Comment (rubidium42) [inline src/newgrf_engine.cpp]: Wouldn't `std::distance(std::begin(common_cargoes), cargo_it)` be conceptually better than `cargo_it - std::begin(common_cargoes)`? After all, if you're using `std::begin(common_cargoes)` over `common_cargoes.begin()`, then you're essentially saying that `common_cargoes` is some type with an iterator, so can you assume it is a contiguous block of memory?
Not sure whether a coding style / guideline / consensus about `std::begin(x)` vs `x.begin()` has been made, though `std::distance(x, y)` vs `y - x` should probably result in a similar result.

**Review by rubidium42 — DISMISSED** [review]
Mostly nitpicking / rabbit holing.

File: src/newgrf_engine.cpp (hunk 2 above)
Comment (PeterN) [reply src/newgrf_engine.cpp]: `std::distance()` works for me. it shows intent.

**Review by rubidium42 — APPROVED** [review]

**Pull request merged**